server/handlers/validation.go

```diff
@@ -0,0 +1,377 @@
+	timestampRole := data.RoleName(data.CanonicalTimestampRole) // just in case
+	for r, update := range roles {
+		if r == snapshotRole || r == timestampRole {
+			continue
+		}
+		m, ok := snap.Signed.Meta[r]
+		if !ok {
+			return fmt.Errorf("snapshot missing metadata for %s", r)
+		}
+		if int64(len(update.Data)) != m.Length {
+			return fmt.Errorf("snapshot has incorrect length for %s", r)
+		}
+
+		if !checkHashes(m, update.Data) {
+			return fmt.Errorf("snapshot has incorrect hashes for %s", r)
+		}
+	}
+	return nil
+}
+
+func checkHashes(meta data.FileMeta, update []byte) bool {
+	for alg, digest := range meta.Hashes {
+		d := utils.DoHash(alg, update)
+		if !bytes.Equal(digest, d) {
+			return false
+		}
+	}
+	return true
+}
+
+func validateTargets(role string, roles map[string]storage.MetaUpdate, kdb *keys.KeyDB) (*data.SignedTargets, error) {
+	// TODO: when delegations are being validated, validate parent
+	//       role exists for any delegation
+	s := &data.Signed{}
+	err := json.Unmarshal(roles[role].Data, s)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse %s", role)
+	}
+	// version specifically gets validated when writing to store to
+	// better handle race conditions there.
+	if err := signed.Verify(s, role, 0, kdb); err != nil {
+		return nil, err
+	}
+	t, err := data.TargetsFromSigned(s)
+	if err != nil {
+		return nil, err
+	}
+	if !data.ValidTUFType(t.Signed.Type, data.CanonicalTargetsRole) {
+		return nil, fmt.Errorf("%s has wrong type", role)
+	}
+	return t, nil
+}
+
+// check the snapshot is present. If it is, the heirarchy
+// of the update is OK. This seems like a simplistic check
+// but is completely sufficient for all possible use cases:
+// 1. the user is updating only the snapshot.
+// 2. the user is updating a targets (incl. delegations) or
+//    root metadata. This requires they also provide a new
+//    snapshot.
+// N.B. users should never be updating timestamps. The server
+//      always handles timestamping. If the user does send a
+//      timestamp, the server will replace it on next
+//      GET timestamp.jsonshould it detect the current
+//      snapshot has a different hash to the one in the timestamp.
+func heirarchyOK(roles map[string]storage.MetaUpdate) error {
+	snapshotRole := data.RoleName(data.CanonicalSnapshotRole)
+	if _, ok := roles[snapshotRole]; !ok {
+		return errors.New("snapshot missing from update")
+	}
+	return nil
+}
+
+func validateRoot(gun string, oldRoot, newRoot []byte) (*data.SignedRoot, error) {
+	var parsedOldRoot *data.SignedRoot
+	parsedNewRoot := &data.SignedRoot{}
+
+	if oldRoot != nil {
+		parsedOldRoot = &data.SignedRoot{}
+		err := json.Unmarshal(oldRoot, parsedOldRoot)
+		if err != nil {
+			// TODO(david): if we can't read the old root should we continue
+			//             here to check new root self referential integrity?
+			//             This would permit recovery of a repo with a corrupted
+			//             root.
```

Comment:
Hmm. My instinct is that we should error out if the old root is corrupted, but I can't think of any obvious attacks that continuing here would open up.